rhino/src/main/java/org/mozilla/javascript/NativeString.java

```diff
@@ -772,6 +780,56 @@ else if (Normalizer.Form.NFC.name().equals(formStr))
 
                         return str.substring(k, k + 1);
                     }
+                case Id_isWellFormed:
+                    {
+                        String str = ScriptRuntime.toString(requireObjectCoercible(cx, thisObj, f));
+                        int len = str.length();
+                        boolean foundLeadingSurrogate = false;
+                        for (int i = 0; i < len; i++) {
+                            char c = str.charAt(i);
+                            if (NativeJSON.isLeadingSurrogate(c)) {
+                                if (foundLeadingSurrogate) {
+                                    return false;
+                                }
+                                foundLeadingSurrogate = true;
+                            } else if (NativeJSON.isTrailingSurrogate(c)) {
+                                if (!foundLeadingSurrogate) {
+                                    return false;
+                                }
+                                foundLeadingSurrogate = false;
+                            } else if (foundLeadingSurrogate) {
+                                return false;
+                            }
+                        }
+                        return !foundLeadingSurrogate;
+                    }
+                case Id_toWellFormed:
+                    {
+                        String str = ScriptRuntime.toString(requireObjectCoercible(cx, thisObj, f));
```

Comment:
Maybe this kind of string processing can be optimized even more. Have a first loop that simply traverses the string until you reach a char that has to be replaced. Then create the string builder an use the already processed chars as initial input. Now you can process like you did.
This ends up with more code but avoidas the string buider and the reconstruction of the string in all standard cases. I assume that most of the json to be normalized does not contain surrogates.

Reply:
That's a great point, thanks! I can try this out and see how it goes.